lint/capability_fields_analyser.go

```diff
@@ -0,0 +1,140 @@
+/*
+ * Cadence-lint - The Cadence linter
+ *
+ * Copyright 2019-2023 Dapper Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package lint
+
+import (
+	"github.com/onflow/cadence/runtime/ast"
+	"github.com/onflow/cadence/runtime/common"
+	"github.com/onflow/cadence/tools/analysis"
+)
+
+func DetectCapabilityType(typeToCheck ast.Type, structWithPubCapabilities map[string]struct{}) bool {
+	const capabilityTypeName = "Capability"
+	switch downcastedType := typeToCheck.(type) {
+	case *ast.NominalType:
+		_, found := structWithPubCapabilities[downcastedType.Identifier.Identifier]
+		return downcastedType.Identifier.Identifier == capabilityTypeName || found
+	case *ast.OptionalType:
+		return DetectCapabilityType(downcastedType.Type, structWithPubCapabilities)
+	case *ast.VariableSizedType:
+		return DetectCapabilityType(downcastedType.Type, structWithPubCapabilities)
+	case *ast.ConstantSizedType:
+		return DetectCapabilityType(downcastedType.Type, structWithPubCapabilities)
+	case *ast.DictionaryType:
+		return DetectCapabilityType(downcastedType.KeyType, structWithPubCapabilities) || DetectCapabilityType(downcastedType.ValueType, structWithPubCapabilities)
+	case *ast.FunctionType:
+		return false
+	case *ast.ReferenceType:
+		return DetectCapabilityType(downcastedType.Type, structWithPubCapabilities)
+	case *ast.RestrictedType:
+		return false
+	case *ast.InstantiationType:
+		return DetectCapabilityType(downcastedType.Type, structWithPubCapabilities)
+	default:
+		panic("Unknown type")
+	}
+}
+
+func CollectStructsWithPublicCapabilities(inspector *ast.Inspector) (map[string]struct{}, map[ast.Identifier]struct{}) {
+	structWithPubCapabilities := make(map[string]struct{})
+	fieldsInStruct := make(map[ast.Identifier]struct{})
+	inspector.Preorder(
+		[]ast.Element{(*ast.CompositeDeclaration)(nil)},
+		func(element ast.Element) {
+			switch declaration := element.(type) {
+			case *ast.CompositeDeclaration:
+				{
+					if declaration.CompositeKind != common.CompositeKindStructure {
```

Comment:
Good question. My thought process was the following ( I mean I saw the idea of looking into structs in the Nozim's patch) -
Structures can be used as types to declare fields in contracts. So, in that way we would have a map of the structs that have a potentially dangerous public capabilities that might leak. I am referring to this test -
```
			`
 		pub contract MyContract {
 		   pub resource Counter {
 		     priv var count: Int
 		
 		     init(count: Int) {
 		       self.count = count
 		     }
 		   }
 		
 		   pub struct ContractData {
 			   pub var owner: Capability
 		       init(cap: Capability) {
 		           self.owner = cap
 		       }
 		   }

			pub struct ContractDataIgnored {
 			   pub var owner: Capability
 		       init(cap: Capability) {
 		           self.owner = cap
 		       }
 		   }
 		
 		   pub var contractData: ContractData
 		
 		   init(){
 		       self.contractData = ContractData(cap:
 		           self.account.getCapability<&Counter>(/public/counter)
 		       )
 		   }
 		}`,
```

I think for contracts it doesn't make sense to have such a map. Or is it? I have my day off on Friday so can take a more in depth look.

For resources you are probably right as a resource might contain a public capability which links to the other resource. I need to check Cadence a bit as it's been a while since I looked into it.

Reply:
That makes sense. Given that a public capability fields is just as problematic in a resource or contract as it is in a struct, it might make sense to just detect it everywhere, i.e. remove the if-statement.